docs/refactor_notes.md

```diff
@@ -0,0 +1,139 @@
+# refactor notes
+
+## FLow Layer
+
+The flow layer needs to be refactored in full, it has to be seperated into smaller more defined components.
+
+This refactor has the following "phases"
+
+1. review relations
+---
+
+* list the current components inside the flow layer
+* list the responsibilities of each component
+* list the relations between each component
+
+---
+2. modify relations
+---
+
+* depulicate aggregate responsibilities over all components
+* re organize responsibilities into distinct "osi stack" like layers
+* re structure components into layers such that they only talk to layers above and below
+
+---
+3. spec out interfaces
+---
+
+* given the new component relations define new compilation targets (liblokinet-whatever) and set up the high level deps
+* in each component define interface types for each responsibility with their relations outside the component itself
+* define internal component relations
+* define unit tests for each responsibility in each component
+
+---
+4. refactor
+---
+
+* rearrange the code from the existing pile into the new structure
+* add unit tests before moving code into place to ensure it does what i think it does
+
+
+## Concrete Plan
+
+---
+1. define new interfaces and structure
+---
+
+* define new cmake libraries that split up lokinet-amalgum more
+* define new interface types in headers in each new split up library
+
+---
+2. wire up old implementation to new interface
+---
+
+* add unit tests that can fail when wired up to old code
+* wire up old code to new interface
+* verify working state
+
+---
+3. add new implementation to new interface
+---
+
+* implement new interface so that unit tests pass with new implementation
+* verify working state
+
+
+
+### new interface types (1)
+
+the component to refactor is `llarp::service::Context`
+
+the types that will get a new interface type are:
+
+`llarp::service::Context` -> `llarp::service::I_ServiceContext`
+
+`llarp::service::Endpoint` -> `llarp::service::I_Endpoint`
```

Comment:
Is there ever going to be more than one implementer of these interfaces?  I feel like we already have too many abstract interfaces that serve little purpose right now because they only have (and likely only ever will have) one single implementation.

Reply:
Additionally, for performance reasons, we ought to avoid unnecessary virtual inheritance, particularly for anything that gets used in hot code: there is a run-time cost (albeit small) of virtual lookup table access.

Reply:
fair